spec/haskell/src/SEL4/Object/Reply.lhs

```diff
@@ -34,51 +34,34 @@ This module specifies the behavior of reply objects.
 >     reply <- getReply replyPtr
 >     setReply replyPtr (upd reply)
 
+> bindScReply :: PPtr SchedContext -> PPtr Reply -> Kernel ()
+> bindScReply scPtr replyPtr = do
+>     sc <- getSchedContext scPtr
+>     scReplyOpt <- return $ scReply sc
+>     when (scReplyOpt /= Nothing) $ do
+>         scReplyPtr <- return $ fromJust scReplyOpt
+>         scReply <- getReply scReplyPtr
+>         setReply scReplyPtr (scReply { replyNext = Just (Next replyPtr) })
+>     reply <- getReply replyPtr
+>     setReply replyPtr (reply { replyPrev = scReplyOpt })
+>     setSchedContext scPtr (sc { scReply = Just replyPtr })
+>     reply <- getReply replyPtr
+>     setReply replyPtr (reply { replyNext = Just (Head $ scPtr) })
```

Comment:
Yes, it is structured this way to make the corres proofs easier but it perhaps makes the wp proofs harder. In this case because the state-relation is highly non-trivial (different models of reply queues) I think it's better to try and make all the different specs have clearly corresponding commands and not to reorder anything. This way we don't have to do so many monadic rewrites or make any complicated arguments about why reorderings are safe. This is just my general preference though, so happy to try another path if there is a need to.

Reply:
We have `updateReply`. I think we should be using it here instead of the `getReply`/`setReply` pair. I agree that we need to do the updates step by step because each of `prev`/`next` updates can have different semantics, and it is a pain having to separate bundled updates depending the cases. But I don’t think there is much reason not to use `updateReply`.

Reply:
Cheers. I'll start making that change.